src/gui_common/AddWindowReorderingSupportToSiblings.cs

```diff
@@ -0,0 +1,328 @@
+            {
+                yield return windowReorderingSupport;
+            }
+        }
+        else
+        {
+            foreach (var windowReorderingSupport in GetWindowReorderingAncestors(startingNode, paths))
+                yield return windowReorderingSupport;
+        }
+    }
+
+    public override void _ExitTree()
+    {
+        windowReorderingAncestors.Clear();
+        connectedWindows.Clear();
+        connectedSiblings.Clear();
+        isReady = false;
+    }
+
+    public void ConnectWindow(CustomWindow window, Node topNode, bool recursive)
+    {
+        // Make sure to be connected to other window reordering nodes before trying to connect the window
+        Setup();
+
+        if (recursive)
+        {
+            // Connection is recursive, connect to other window reordering nodes
+            for (int i = 0; i < windowReorderingAncestors.Count; i++)
+            {
+                windowReorderingAncestors[i].Node.ConnectWindow(
+                    window, windowReorderingAncestors[i].Sibling, recursive);
+            }
+        }
+
+        window.Connect(nameof(Dragged), this, nameof(OnWindowReorder));
+
+        connectedWindows.Add(window, topNode);
+        connectedSiblings.Add(topNode);
+
+        // Update top sibling
+        if (topSibling is null || topSibling.GetIndex() < topNode.GetIndex())
+            topSibling = topNode;
+    }
+
+    public void DisconnectWindow(CustomWindow window, bool recursive)
+    {
+        if (recursive)
+        {
+            // Connection was recursive, disconnect from other window reordering nodes
+            for (int i = 0; i < windowReorderingAncestors.Count; i++)
+            {
+                windowReorderingAncestors[i].Node.DisconnectWindow(window, recursive);
+            }
+        }
+
+        window.Disconnect(nameof(Dragged), this, nameof(OnWindowReorder));
+
+        var windowSibling = connectedWindows[window];
+        connectedWindows.Remove(window);
+
+        // Find if another connected window has the same sibling
+        bool foundSibling = false;
+        foreach (var sibling in connectedSiblings)
+        {
+            if (sibling == windowSibling)
+            {
+                foundSibling = true;
+                break;
+            }
+        }
+
+        if (!foundSibling)
+        {
+            // No other window has the same sibling so it can be removed
+            connectedSiblings.Remove(windowSibling);
+        }
+    }
+
+    /// <summary>
+    ///   Finds whose windows ancestor is currently on the top
+    /// </summary>
+    public void UpdateTopSibling()
+    {
+        if (topSibling is not null)
+        {
+            // Top sibling is already current
+            return;
+        }
+
+        var siblings = parent.GetChildren();
+
+        // Search for the first sibling that is part of the reordering system
+        for (int i = siblings.Count - 1; i >= 0; i--)
+        {
+            var sibling = (Node)siblings[i];
```

Comment:
```suggestion
        // Search for the first sibling that is part of the reordering system
        foreach (Node sibling in parent.GetChildren())
        {
```

Reply:
Sorry I misread the loop order, I see you reversed the loop.

I saw below that UpdateTopSibling seems to be called quite often in a situation where the last child is already on top.

So actually I suggest you do something like:
```
var childCount = parent.GetChildCount();

for (int i = childCount - 1; ...)
{
    var sibling = parent.GetChild(i);
}
```

That way a full list of children don't need t obe generated and marshalled into C# unnecessarily.

Reply:
This change is good.
It's called often but most of the time topSibling is not null. This part of the code is used only when topSibling becomes invalid (like after disconnecting a window that is on the top).